Sources/NIOCore/ByteBuffer-binaryEncodedLengthPrefix.swift

```diff
@@ -0,0 +1,313 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2024 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if os(Windows)
+import ucrt
+#elseif canImport(Darwin)
+import Darwin
+#elseif canImport(Glibc)
+import Glibc
+#elseif canImport(Musl)
+import Musl
+#elseif canImport(Bionic)
+import Bionic
+#else
+#error("The Byte Buffer module was unable to identify your C library.")
+#endif
+
+/// Describes a way to encode and decode an integer as bytes
+///
+public protocol BinaryIntegerEncodingStrategy {
+    /// Read an integer from a buffer. The reader index should be moved accordingly
+    /// - Returns: The integer
+    func readInteger<IntegerType: FixedWidthInteger>(
+        as: IntegerType.Type,
+        from buffer: inout ByteBuffer
+    ) -> IntegerType?
+
+    /// Write an integer to a buffer. The writer index should be moved accordingly
+    func writeInteger<IntegerType: FixedWidthInteger>(
+        _ integer: IntegerType,
+        to buffer: inout ByteBuffer
+    ) -> Int
+
+    /// When writing an integer using this strategy, how many bytes we should reserve
```

Comment:
I found this and `writeIntegerWithReservedSpace` quite unintuitive without looking at how they were used. It became clearer when I started reading "Integer" as "Length". Is that the right way to think about it?

Reply:
Yeah i agree this is confusing

The idea here is that `reservedSpaceForInteger` is the amount of space that someone should reserve if theyre intending to write an integer using your strategy
And the `writeIntegerWithReservedSpace` is for when someone has already reserved some space, and now wants you to write an integer into that space. You can then make the decision on whether you want to try and use exactly that much space (this is what the quic impl does) even if you could've used less. Ie you choose whether you optimize for the smallest possible encoding of the int, or for trying to avoid moving bytes around

This is in practice currently only used by the writePrefixed* functions so yes, there the integer is a length. But i don't think it _has_ to be a length

Reply:
It _can_ be used to write an integer, but in the context of this protocol _should_ it be used for anything other than the length? 

IMO we should make the intent of the requirement clear, if you wanted to write an integer then you should just call the `writeInteger(_:to:)` method.